***MGSC 662 - Assignment 1***

***Fresh Milk Transportation***

*BalancedMilk* is an old dairy-transportation firm renowned for the freshness of the milk it distributes. It has long-term fixed-amount contracts with eight dairy farms (suppliers) and distributes their production among ten demand markets. Transporting milk for a long time from suppliers to demand markets, BalancedMilk is well aware of the milk demand in each demand market and the transportation cost from each supplier to each demand center.

In the following table, the number in front of each supply and demand center represents the supply and demand (in tonnes.) Also, for each supply-demand pair, the transportation cost ($ per tonne) is depicted.

| | $D_1$(90)| $D_2$(100) | $D_3$(150) | $D_4$(190) | $D_5$(180) | $D_6$(240) | $D_7$(210) | $D_8$(90) | $D_9$(160) | $D_{10}$(70) |
|--| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
| $S_1$(110) |20|49|16|30|8|35|21|40|10|12|
| $S_2$(80) |15|53|7|20|47|11|16|17|15|44|
| $S_3$(100) |22|25|42|22|31|9|11|29|20|5|
| $S_4$(240) |45|6|33|35|49|25|30|47|32|31|
| $S_5$(100) |9|12|41|15|38|14|53|22|12|13|
| $S_6$(280) |21|24|32|49|5|47|30|23|37|8|
| $S_7$(130) |12|19|5|28|47|39|15|35|9|51|
| $S_8$(440) |34|17|10|21|9|33|14|26|19|45|

Keep the following points in mind when solving the problems:

**Note 1**: The firm should not supply milk to a demand market **more** than its demand, i.e., should not **over-supply** it.

**Note 2**: Write your code so that your answer to each part is independent of the other. For instance, to get the solution for Problem 2, one should not be forced to run Problem 1 beforehand. 

For each of the following strategies, formulate them as a LP problem and solve them using Gurobi:

**Problem 1:**

The firm wants to distribute all the supplied milk as efficiently as possible. Formulate the problem of minimizing the total transportation cost using Gurobi and solve it. Then, report the optimal allocation and the minimum cost.

In [1]:
Cost = [[20, 49, 16, 30,  8, 35, 21, 40, 10, 12],
        [15, 53, 7, 20, 47, 11, 16, 17, 15, 44],
        [22, 25, 42, 22, 31, 9, 11, 29, 20, 5],
        [45,  6, 33, 35, 49, 25, 30, 47, 32, 31],
        [9, 12, 41, 15, 38, 14, 53, 22, 12, 13],
        [21, 24, 32, 49, 5, 47, 30, 23, 37, 8],
        [12, 19,  5, 28, 47, 39, 15, 35, 9, 51],
        [34, 17, 10, 21, 9, 33, 14, 26, 19, 45]]

SupplyCenter = ['S_1', 'S_2', 'S_3', 'S_4', 'S_5', 'S_6', 'S_7', 'S_8']
Supply = [110, 80, 100, 240, 100, 280, 130, 440]
DemandMarket = ['D_1', 'D_2', 'D_3', 'D_4', 'D_5', 'D_6', 'D_7', 'D_8', 'D_9', 'D_10']
Demand = [90, 100, 150, 190, 180, 240, 210, 90, 160, 70]

# Your solution for Problem 1:

In [2]:
import gurobipy as gb
from gurobipy import *

In [3]:
prob = gb.Model("Fresh Milk Transportation")

Set parameter Username
Academic license - for non-commercial use only - expires 2023-09-10


In [4]:
S = len(Supply)
D = len(Demand)

Adding variables:

In [5]:
X = prob.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = ["X_"+i+"_"+j for i in SupplyCenter for j in DemandMarket])

Objective function:

In [6]:
prob.setObjective(sum(Cost[i][j]*X[i,j] for i in range(S) for j in range(D)), GRB.MINIMIZE)

Adding suppy constraints:

In [7]:
for i in range(S):
    prob.addConstr(sum(X[i,j] for j in range(D))<=Supply[i])

Adding demand constraints:

In [8]:
for j in range(D):
    prob.addConstr(sum(X[i,j] for i in range(S)) >= Demand[j])

In [9]:
prob.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 18 rows, 80 columns and 160 nonzeros
Model fingerprint: 0xcc3efc64
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e+01, 4e+02]
Presolve time: 0.02s
Presolved: 18 rows, 80 columns, 160 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.480000e+03   0.000000e+00      0s
      23    1.8440000e+04   0.000000e+00   0.000000e+00      0s

Solved in 23 iterations and 0.04 seconds (0.00 work units)
Optimal objective  1.844000000e+04


In [10]:
for v in prob.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))

X_S_1_D_9 = 110
X_S_2_D_3 = 20
X_S_2_D_8 = 60
X_S_3_D_6 = 100
X_S_4_D_2 = 100
X_S_4_D_6 = 140
X_S_5_D_1 = 90
X_S_5_D_4 = 10
X_S_6_D_5 = 180
X_S_6_D_8 = 30
X_S_6_D_10 = 70
X_S_7_D_3 = 80
X_S_7_D_9 = 50
X_S_8_D_3 = 50
X_S_8_D_4 = 180
X_S_8_D_7 = 210


In [11]:
prob.objVal

18440.0

Due to a cattle virus outbreak in the largest supplier ($S_8$), the health ministry shut down this dairy farm. Left with seven suppliers and unchanged demands, BalancedMilk is facing a supply shortage. Therefore, the management establishes two teams to search for short-term and long-term strategies to combat this issue. While the *long-term team* is looking for alternative suppliers or new contracts, we want to evaluate the strategies offered by the *short-term team* in the following problems.

**Problem 2:**

Suppose not supplying milk to a demand market is costless. Nevertheless, the firm should supply all the milk the remaining seven supply centers provide. Also, remember that the firm cannot oversupply a demand market. Formulate the problem of minimizing the total transportation cost using Gurobi and solve it. Then, report the optimal allocation and the minimum cost. 

(Hint: one way of modeling this situation is to replace $S_8$ with a *dummy supplier* that provides the same amount of milk previously provided by $S_8$ but at zero cost.)


In [12]:
# Your solution for Problem 2:

In [13]:
prob_1 = gb.Model("Fresh Milk Transportation Without S8")

In [14]:
Cost_new = [[20, 49, 16, 30,  8, 35, 21, 40, 10, 12],
        [15, 53, 7, 20, 47, 11, 16, 17, 15, 44],
        [22, 25, 42, 22, 31, 9, 11, 29, 20, 5],
        [45,  6, 33, 35, 49, 25, 30, 47, 32, 31],
        [9, 12, 41, 15, 38, 14, 53, 22, 12, 13],
        [21, 24, 32, 49, 5, 47, 30, 23, 37, 8],
        [12, 19,  5, 28, 47, 39, 15, 35, 9, 51],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]

In [15]:
X = prob_1.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = ["X_"+i+"_"+j for i in SupplyCenter for j in DemandMarket])

In [16]:
prob_1.setObjective(sum(Cost_new[i][j]*X[i,j] for i in range(S) for j in range(D)), GRB.MINIMIZE)

In [17]:
for i in range(S):
    prob_1.addConstr(sum(X[i,j] for j in range(D))<=Supply[i])

In [18]:
for j in range(D):
    prob_1.addConstr(sum(X[i,j] for i in range(S)) >= Demand[j])

In [19]:
prob_1.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 18 rows, 80 columns and 160 nonzeros
Model fingerprint: 0x2e5e0acc
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e+01, 4e+02]
Presolve time: 0.01s
Presolved: 18 rows, 80 columns, 160 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.480000e+03   0.000000e+00      0s
      22    1.0670000e+04   0.000000e+00   0.000000e+00      0s

Solved in 22 iterations and 0.02 seconds (0.00 work units)
Optimal objective  1.067000000e+04


In [20]:
for v in prob_1.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))

X_S_1_D_9 = 110
X_S_2_D_3 = 60
X_S_2_D_6 = 20
X_S_3_D_6 = 80
X_S_3_D_7 = 20
X_S_4_D_2 = 100
X_S_4_D_6 = 140
X_S_5_D_1 = 90
X_S_5_D_9 = 10
X_S_6_D_5 = 180
X_S_6_D_8 = 30
X_S_6_D_10 = 70
X_S_7_D_3 = 90
X_S_7_D_9 = 40
X_S_8_D_4 = 190
X_S_8_D_7 = 190
X_S_8_D_8 = 60


In [21]:
prob_1.objVal

10670.0

#### Above results explained:
S_8 is considered as a dummy supplier, which means that everything supplied from it is fictional.

* D_4 needs 190 tonnes of milk, which are supposedly supplied by S_8. In reality, D_4 is getting nothing as its only supplier is S_8.
* D_7 needs 210 tonnes of milk. 190 tonnes are supposedly supplied by S_8, and 20 tonnes by S_3. In reality, D_7 is only getting the 20 tonnes of milk from S_3 and nothing from S_8.
* D_8 needs 90 tonnes of milk. 60 tonnes are supposedly supplied by S_8, and 30 tonnes by S_6. In reality, D_8 is only getting the 30 tonnes of milk from S_6 and nothing from S_8.

(Btw: 190 + 190 + 60 = 440, which is the total amount of milk initially supplied by S_8)

Thus, we are under supplying for D_4, D_7 and D_8.

Moreover, total supply excluding S_8 = 1040 tonnes of milk, which is equal to the amount of actually supplied milk as seen above. Thus, all available milk has been supplied in this model. 

Finally, the optimal allocation and minimal cost are the following: 
##### Allocation: 
X_S_1_D_9 = 110
X_S_2_D_3 = 60
X_S_2_D_6 = 20
X_S_3_D_6 = 80
X_S_3_D_7 = 20
X_S_4_D_2 = 100
X_S_4_D_6 = 140
X_S_5_D_1 = 90
X_S_5_D_9 = 10
X_S_6_D_5 = 180
X_S_6_D_8 = 30
X_S_6_D_10 = 70
X_S_7_D_3 = 90
X_S_7_D_9 = 40
##### Minimal cost:
$10,670.00

**Problem 3:**

In the real world, with fierce competitors ready to attack *BalancedMilk*'s market share, the cost of not supplying milk to a demand center is not zero. Additionally, like Problem 2, the firm should supply all the milk provided by the remaining seven supply centers. Suppose the cost of not delivering each tonne of milk to each demand center is  20\\$/tone. Formulate the problem of minimizing the sum of total transportation and not supplying costs using Gurobi and solve it. Then, report the optimal allocation and the minimum cost. Now suppose instead of 20\\$/tone, the not supplying cost is 100\$/tone. Again, formulate the problem of minimizing the sum of total transportation and not supplying costs using Gurobi and solve it. Then, report the optimal allocation and the minimum cost. 

Can you detect any patterns in the optimal allocation in this problem and the previous one? How about the difference between the minimum costs? Explain your intuition.

In [22]:
# Your solution for Problem 3:

#### Cost of not supplying: $20/tonne

In [23]:
prob_2 = gb.Model("Fresh Milk Transportation Without S8 costs $20/tonne")

In [24]:
Cost_2 = [[20, 49, 16, 30,  8, 35, 21, 40, 10, 12],
        [15, 53, 7, 20, 47, 11, 16, 17, 15, 44],
        [22, 25, 42, 22, 31, 9, 11, 29, 20, 5],
        [45,  6, 33, 35, 49, 25, 30, 47, 32, 31],
        [9, 12, 41, 15, 38, 14, 53, 22, 12, 13],
        [21, 24, 32, 49, 5, 47, 30, 23, 37, 8],
        [12, 19,  5, 28, 47, 39, 15, 35, 9, 51],
        [20, 20, 20, 20, 20, 20, 20, 20, 20, 20]]

In [25]:
Y = prob_2.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = ["X_"+i+"_"+j for i in SupplyCenter for j in DemandMarket])

In [26]:
prob_2.setObjective(sum(Cost_2[i][j]*Y[i,j] for i in range(S) for j in range(D)), GRB.MINIMIZE)

In [27]:
for i in range(S):
    prob_2.addConstr(sum(Y[i,j] for j in range(D))<=Supply[i])

In [28]:
for j in range(D):
    prob_2.addConstr(sum(Y[i,j] for i in range(S)) >= Demand[j])

In [29]:
prob_2.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 18 rows, 80 columns and 160 nonzeros
Model fingerprint: 0x6d48932d
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e+01, 4e+02]
Presolve time: 0.02s
Presolved: 18 rows, 80 columns, 160 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.480000e+03   0.000000e+00      0s
      23    1.9470000e+04   0.000000e+00   0.000000e+00      0s

Solved in 23 iterations and 0.03 seconds (0.00 work units)
Optimal objective  1.947000000e+04


In [30]:
for v in prob_2.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))

X_S_1_D_9 = 110
X_S_2_D_3 = 60
X_S_2_D_6 = 20
X_S_3_D_6 = 80
X_S_3_D_7 = 20
X_S_4_D_2 = 100
X_S_4_D_6 = 140
X_S_5_D_1 = 90
X_S_5_D_9 = 10
X_S_6_D_5 = 180
X_S_6_D_8 = 30
X_S_6_D_10 = 70
X_S_7_D_3 = 90
X_S_7_D_9 = 40
X_S_8_D_4 = 190
X_S_8_D_7 = 190
X_S_8_D_8 = 60


In [31]:
prob_2.objVal

19470.0

When the cost of not supplying is 20 dollars/tonne of milk, the optimized cost of transportation is equal to 19470 dollars, which is equal to the optimal cost of transportation found when cost of not supplying was equal to 0, to which we add the amount of milk that was supposed to be supplied by S_8 times 20 dollars:
* (190+190+60) x 20 = 8800
* 10,670.00 + 8,800.00 = 19,470

Regarding the optimal allocation it is still the same as the one we got before, when cost of not supplying was null.

#### Cost of not supplying: $100/tonne

In [32]:
prob_3 = gb.Model("Fresh Milk Transportation Without S8 costs $100/tonne")

In [33]:
Cost_3 = [[20, 49, 16, 30,  8, 35, 21, 40, 10, 12],
        [15, 53, 7, 20, 47, 11, 16, 17, 15, 44],
        [22, 25, 42, 22, 31, 9, 11, 29, 20, 5],
        [45,  6, 33, 35, 49, 25, 30, 47, 32, 31],
        [9, 12, 41, 15, 38, 14, 53, 22, 12, 13],
        [21, 24, 32, 49, 5, 47, 30, 23, 37, 8],
        [12, 19,  5, 28, 47, 39, 15, 35, 9, 51],
        [100, 100, 100, 100, 100, 100, 100, 100, 100, 100]]

In [34]:
Z = prob_3.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = ["X_"+i+"_"+j for i in SupplyCenter for j in DemandMarket])

In [35]:
prob_3.setObjective(sum(Cost_3[i][j]*Z[i,j] for i in range(S) for j in range(D)), GRB.MINIMIZE)

In [36]:
for i in range(S):
    prob_3.addConstr(sum(Z[i,j] for j in range(D))<=Supply[i])

In [37]:
for j in range(D):
    prob_3.addConstr(sum(Z[i,j] for i in range(S)) >= Demand[j])

In [38]:
prob_3.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 18 rows, 80 columns and 160 nonzeros
Model fingerprint: 0x09c330f5
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 1e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e+01, 4e+02]
Presolve time: 0.01s
Presolved: 18 rows, 80 columns, 160 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.480000e+03   0.000000e+00      0s
      24    5.4670000e+04   0.000000e+00   0.000000e+00      0s

Solved in 24 iterations and 0.02 seconds (0.00 work units)
Optimal objective  5.467000000e+04


In [39]:
for v in prob_3.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))

X_S_1_D_9 = 110
X_S_2_D_3 = 60
X_S_2_D_6 = 20
X_S_3_D_6 = 80
X_S_3_D_7 = 20
X_S_4_D_2 = 100
X_S_4_D_6 = 140
X_S_5_D_1 = 90
X_S_5_D_9 = 10
X_S_6_D_5 = 180
X_S_6_D_8 = 30
X_S_6_D_10 = 70
X_S_7_D_3 = 90
X_S_7_D_9 = 40
X_S_8_D_4 = 190
X_S_8_D_7 = 190
X_S_8_D_8 = 60


In [40]:
prob_3.objVal

54670.0

Same again happens for cost of not supplying = 100 dollars/tonne:

cost of total transportation = cost of total trasnsportation with no cost of not supplying + 440 x 100 = 10670 + 44000 = 54670
                           
Again, optimal allocation of supply to demand stays the same

**Problem 4:**

In Problems 2 and 3, we assumed the firm is committed to distributing all the milk supplied by the seven remaining supply centers. Suppose the firm does not entertain this constraint anymore, and the cost of not supplying each tonne of milk to each demand center is  20\\$/tone. Formulate the problem of minimizing the sum of total transportation and not supplying costs using Gurobi and solve it. Then, report the optimal allocation and the minimum cost. Now suppose instead of 20\\$/tone, the not supplying cost is 100\$/tone. Again, formulate the problem of minimizing the sum of total transportation and not supplying costs using Gurobi and solve it. Then, report the optimal allocation and the minimum cost. 

Do you see any difference between the optimal allocation in this problem and Problem 3? Do you have any educated guesses on why it is the case?

In [41]:
# Your solution for Problem 4:

prob_4 = gb.Model("Problem 4 part 1")

In [42]:
W = prob_4.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = ["X_"+i+"_"+j for i in SupplyCenter for j in DemandMarket])

In [43]:
prob_4.setObjective(sum(Cost_2[i][j]*W[i,j] for i in range(S) for j in range(D)), GRB.MINIMIZE)
# cost_2 is the cost matrix for when the S_8 row is equal to 20 dollars

In [44]:
for i in range(S-1):
    prob_4.addConstr(sum(W[i,j] for j in range(D))<=Supply[i])

for j in range(D):
    prob_4.addConstr(sum(W[i,j] for i in range(S)) >= Demand[j])

prob_4.addConstr(sum(W[7,j] for j in range(D))<=1480)

<gurobi.Constr *Awaiting Model Update*>

In [45]:
prob_4.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 18 rows, 80 columns and 160 nonzeros
Model fingerprint: 0x439e2a8f
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e+01, 1e+03]
Presolve time: 0.01s
Presolved: 18 rows, 80 columns, 160 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.480000e+03   0.000000e+00      0s
      23    1.8640000e+04   0.000000e+00   0.000000e+00      0s

Solved in 23 iterations and 0.02 seconds (0.00 work units)
Optimal objective  1.864000000e+04


In [46]:
for v in prob_4.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))

X_S_1_D_9 = 110
X_S_2_D_3 = 20
X_S_2_D_6 = 60
X_S_3_D_6 = 100
X_S_4_D_2 = 100
X_S_5_D_1 = 90
X_S_5_D_9 = 10
X_S_6_D_5 = 180
X_S_6_D_10 = 70
X_S_7_D_3 = 130
X_S_8_D_4 = 190
X_S_8_D_6 = 80
X_S_8_D_7 = 210
X_S_8_D_8 = 90
X_S_8_D_9 = 40


In [47]:
prob_4.objVal

18640.0

In [48]:
prob_5 = gb.Model("Problem 4 part 2")

In [49]:
T = prob_5.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = ["X_"+i+"_"+j for i in SupplyCenter for j in DemandMarket])

In [50]:
prob_5.setObjective(sum(Cost_3[i][j]*T[i,j] for i in range(S) for j in range(D)), GRB.MINIMIZE)
# cost_3 is the cost matrix for when the S_8 row is equal to 100 dollars

In [51]:
for i in range(S-1):
    prob_5.addConstr(sum(T[i,j] for j in range(D))<=Supply[i])

for j in range(D):
    prob_5.addConstr(sum(T[i,j] for i in range(S)) >= Demand[j])

prob_5.addConstr(sum(T[7,j] for j in range(D))<=1480)

<gurobi.Constr *Awaiting Model Update*>

In [52]:
prob_5.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 18 rows, 80 columns and 160 nonzeros
Model fingerprint: 0x7dca54d5
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 1e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e+01, 1e+03]
Presolve time: 0.01s
Presolved: 18 rows, 80 columns, 160 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.480000e+03   0.000000e+00      0s
      24    5.4670000e+04   0.000000e+00   0.000000e+00      0s

Solved in 24 iterations and 0.03 seconds (0.00 work units)
Optimal objective  5.467000000e+04


In [53]:
for v in prob_5.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))

X_S_1_D_9 = 110
X_S_2_D_3 = 60
X_S_2_D_6 = 20
X_S_3_D_6 = 80
X_S_3_D_7 = 20
X_S_4_D_2 = 100
X_S_4_D_6 = 140
X_S_5_D_1 = 90
X_S_5_D_9 = 10
X_S_6_D_5 = 180
X_S_6_D_8 = 30
X_S_6_D_10 = 70
X_S_7_D_3 = 90
X_S_7_D_9 = 40
X_S_8_D_4 = 190
X_S_8_D_7 = 190
X_S_8_D_8 = 60


In [54]:
prob_5.objVal

54670.0

#### Interpretation of above results:
When we set the price of not supplying to be at 20 dollars per tonne of milk, we get an optimal transportation cost of 18,640.00 dollars, with a supply allocation as shown above. However, when that same not-supplying cost is at 100 dollars, our optimal cost function takes the value of 54,670.00 dollars, same value as in problem 3 where supplying the whole stock was a necessity. 
This means that when cost of not supplying is at 20 dollars, it is cheaper for a our company to not deliver milk. But when it's at 100 dollars, it is more expensive to not deliver milk as their sale price is higher than delivery. Therefore, they would prefer delivering their whole stock and we'd be back to our previous scenario.

The above situation is related to Shadow price.

Shadow price of a constraint: change in optimal objective function value if the constraint’s right-hand side is increased by one unit and all other problem data remains fixed.



**Problem 5:**

In Problems 2, 3 and 4, we assumed the no-supply cost is the same for all demand centers. However, this may not be the case. The estimations by the *short-term team* indicate that the cost of not supplying a tonne of milk to each demand center is as follows:

| | $D_1$| $D_2$ | $D_3$ | $D_4$ | $D_5$ | $D_6$ | $D_7$ | $D_8$ | $D_9$ | $D_{10}$|
|--| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
| Cost of not supplying (\\$/tonne) |15|20|25|30|28|35|32|15|25|10|

Additionally, like Problem 2, the firm should supply all the milk provided by the remaining seven supply centers. Formulate the problem of minimizing the sum of total transportation and not supplying costs using Gurobi and then solve it. Report the optimal allocation and the minimum cost. Is the optimal allocation the same as Problem 2? Why?

In [55]:
# Your solution for Problem 5:
C = [[20, 49, 16, 30,  8, 35, 21, 40, 10, 12],
        [15, 53, 7, 20, 47, 11, 16, 17, 15, 44],
        [22, 25, 42, 22, 31, 9, 11, 29, 20, 5],
        [45,  6, 33, 35, 49, 25, 30, 47, 32, 31],
        [9, 12, 41, 15, 38, 14, 53, 22, 12, 13],
        [21, 24, 32, 49, 5, 47, 30, 23, 37, 8],
        [12, 19,  5, 28, 47, 39, 15, 35, 9, 51],
        [34, 17, 10, 21, 9, 33, 14, 26, 19, 45],
        [15, 20, 25, 30, 28, 35, 32, 15, 25, 10]]
DummySupplier = [110, 80, 100, 240, 100, 280, 130, 0, 440]
DummySupplyCenter = ['S_1', 'S_2', 'S_3', 'S_4', 'S_5', 'S_6', 'S_7', 'S_8', 'S_8_Dummy']
SD=len(DummySupplyCenter)

In [56]:
prob_6 = gb.Model("Problem 5")

In [57]:
Q = prob_6.addVars(SD,D, lb = 0, vtype = GRB.CONTINUOUS, name = ["X_"+i+"_"+j for i in DummySupplyCenter for j in DemandMarket])

In [58]:
prob_6.setObjective(sum(C[i][j]*Q[i,j] for i in range(SD) for j in range(D)), GRB.MINIMIZE)

In [59]:
for i in range(SD):
    prob_6.addConstr(sum(Q[i,j] for j in range(D))<=DummySupplier[i])

for j in range(D):
    prob_6.addConstr(sum(Q[i,j] for i in range(SD)) >= Demand[j])

In [60]:
prob_6.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 19 rows, 90 columns and 180 nonzeros
Model fingerprint: 0x012601c2
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e+01, 4e+02]
Presolve removed 1 rows and 10 columns
Presolve time: 0.01s
Presolved: 18 rows, 80 columns, 160 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.850000e+02   0.000000e+00      0s
      33    2.1980000e+04   0.000000e+00   0.000000e+00      0s

Solved in 33 iterations and 0.03 seconds (0.00 work units)
Optimal objective  2.198000000e+04


In [61]:
for v in prob_6.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))

X_S_1_D_9 = 110
X_S_2_D_6 = 80
X_S_3_D_7 = 100
X_S_4_D_2 = 100
X_S_4_D_6 = 140
X_S_5_D_4 = 80
X_S_5_D_6 = 20
X_S_6_D_5 = 180
X_S_6_D_7 = 100
X_S_7_D_3 = 130
X_S_8_Dummy_D_1 = 90
X_S_8_Dummy_D_3 = 20
X_S_8_Dummy_D_4 = 110
X_S_8_Dummy_D_7 = 10
X_S_8_Dummy_D_8 = 90
X_S_8_Dummy_D_9 = 50
X_S_8_Dummy_D_10 = 70


In [62]:
prob_6.objVal

21980.0

The optimal allocation is different from that in problem 2. Before, the cost of not supplying was always set to 0$. Now, depending on each demand center, the cost changes. Therefore, the model needs to figure out a way of re-allocating the centers to which we do not deliver, again in order to minimize total transporation cost.

**Problem 6:**

Again, suppose not supplying milk to a demand market is costless (like Problem 2,) but the firm is still committed to supplying all the available milk. This time, the firm decides to supply the milk so that each demand center receives at least 50% of its demand. Formulate the problem of minimizing the total transportation costs using Gurobi and solve it. Then, report the optimal allocation and the minimum cost. 

Compare the optimal cost with your answer to Problem 2. Is it more or less? Why?

Suppose the firm sets this bar to 90% - each demand center should receive at least 90% of its demand. Is it possible? What is the maximum possible percentage?

In [63]:
# Your solution for Problem 6:

In [64]:
C2 = [[20, 49, 16, 30,  8, 35, 21, 40, 10, 12],
      [15, 53, 7, 20, 47, 11, 16, 17, 15, 44],
      [22, 25, 42, 22, 31, 9, 11, 29, 20, 5],
      [45,  6, 33, 35, 49, 25, 30, 47, 32, 31],
      [9, 12, 41, 15, 38, 14, 53, 22, 12, 13],
      [21, 24, 32, 49, 5, 47, 30, 23, 37, 8],
      [12, 19,  5, 28, 47, 39, 15, 35, 9, 51]]

SupplyCenter = ['S_1', 'S_2', 'S_3', 'S_4', 'S_5', 'S_6', 'S_7']
Supply = [110, 80, 100, 240, 100, 280, 130]
DemandMarket = ['D_1', 'D_2', 'D_3', 'D_4', 'D_5', 'D_6', 'D_7', 'D_8', 'D_9', 'D_10']
Demand = [90, 100, 150, 190, 180, 240, 210, 90, 160, 70]

S= len(SupplyCenter)
D= len(DemandMarket)

prob_7 = gb.Model("Fresh Milk Transportation with (50%")
H = prob_7.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = [SupplyCenter[i]+"_to_"+DemandMarket[j] for i in range(S) for j in range(D)])

prob_7.setObjective(sum(C2[i][j]*H[i,j] for i in range(S) for j in range(D))+ 0*(sum(Demand[j] for j in range(D)) - sum(H[i,j] for i in range(S) for j in range(D))),
                  GRB.MINIMIZE)

for i in range(S):
    prob_7.addConstr(sum(H[i,j] for j in range(D))==Supply[i])

for j in range(D):
    prob_7.addConstr(sum(H[i,j] for i in range(S))<=Demand[j])    
    
for j in range(D):
    prob_7.addConstr(sum(H[i,j] for i in range(S)) >=Demand[j]*0.5)
    
prob_7.optimize()

print("\nThe optimal allocation is:")
for v in prob_7.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))
        
print("\nThe minimum cost for this problem is: $",prob_7.Objval) 

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 27 rows, 70 columns and 210 nonzeros
Model fingerprint: 0x9102bf9e
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [4e+01, 3e+02]
Presolve removed 10 rows and 0 columns
Presolve time: 0.01s
Presolved: 17 rows, 80 columns, 150 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   2.225000e+02   0.000000e+00      0s
      22    1.1605000e+04   0.000000e+00   0.000000e+00      0s

Solved in 22 iterations and 0.03 seconds (0.00 work units)
Optimal objective  1.160500000e+04

The optimal allocation is:
S_1_to_D_9 = 110
S_2_to_D_3 = 65
S_2_to_D_8 = 15
S_3_to_D_7 = 100
S_4_to_D_2 = 100
S_4_to_D_6 = 135
S_4_to_D_7 = 5
S_5_to_D_1 = 5
S_5_to_D_4 = 95
S_6_to_D_5 = 180
S_6_to_D_8 = 30
S_6_to_D_10 = 7

We can see that when we have a requirement 50%, the transporattion cost is higher than in Question 2. The reason is because this requirement contrains the model to deliver at least 50%, which was nor the case in Q2 where undersupply could be done to any degree.

In [65]:
C2 = [[20, 49, 16, 30,  8, 35, 21, 40, 10, 12],
      [15, 53, 7, 20, 47, 11, 16, 17, 15, 44],
      [22, 25, 42, 22, 31, 9, 11, 29, 20, 5],
      [45,  6, 33, 35, 49, 25, 30, 47, 32, 31],
      [9, 12, 41, 15, 38, 14, 53, 22, 12, 13],
      [21, 24, 32, 49, 5, 47, 30, 23, 37, 8],
      [12, 19,  5, 28, 47, 39, 15, 35, 9, 51]]

SupplyCenter = ['S_1', 'S_2', 'S_3', 'S_4', 'S_5', 'S_6', 'S_7']
Supply = [110, 80, 100, 240, 100, 280, 130]
DemandMarket = ['D_1', 'D_2', 'D_3', 'D_4', 'D_5', 'D_6', 'D_7', 'D_8', 'D_9', 'D_10']
Demand = [90, 100, 150, 190, 180, 240, 210, 90, 160, 70]

S= len(SupplyCenter)
D= len(DemandMarket)

prob_8 = gb.Model("Fresh Milk Transportation with (50%")
E = prob_8.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = [SupplyCenter[i]+"_to_"+DemandMarket[j] for i in range(S) for j in range(D)])

prob_8.setObjective(sum(C2[i][j]*E[i,j] for i in range(S) for j in range(D))+ 0*(sum(Demand[j] for j in range(D)) - sum(E[i,j] for i in range(S) for j in range(D))),
                  GRB.MINIMIZE)

for i in range(S):
    prob_8.addConstr(sum(E[i,j] for j in range(D))==Supply[i])

for j in range(D):
    prob_8.addConstr(sum(E[i,j] for i in range(S))<=Demand[j])    
    
for j in range(D):
    prob_8.addConstr(sum(E[i,j] for i in range(S)) >=Demand[j]*0.9)
    
prob_8.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 27 rows, 70 columns and 210 nonzeros
Model fingerprint: 0xe19f8cf7
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e+00, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [6e+01, 3e+02]
Presolve removed 10 rows and 0 columns
Presolve time: 0.01s
Presolved: 17 rows, 80 columns, 150 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   2.965000e+02   0.000000e+00      0s

Solved in 15 iterations and 0.03 seconds (0.00 work units)
Infeasible model


We can see that for 90%, the problem is not feasable. That is because the maximal percentage of delivery that can be constrained of the model is is (1040/1480) = 70.27% This is obtained by doing the ratio of available stock over total stock. Thus, 90% delivery is an infeasable model.

**Problem 7:**

Suppose the cost of not supplying each tonne of milk to each demand center is 70\\$/tone. As a short-term solution, the firm can ask one of the seven remaining suppliers to add 100 tonnes of milk to its supplying capacity by paying the selected supplier 1000\\$. Which supplier should the firm ask for this capacity boost to minimize costs? What is the change in the costs due to this boost? Can you solve this problem without redefining the problem for each supplier?

In [66]:
# Your solution for Problem 7:

C3 = [[20, 49, 16, 30,  8, 35, 21, 40, 10, 12],
      [15, 53, 7, 20, 47, 11, 16, 17, 15, 44],
      [22, 25, 42, 22, 31, 9, 11, 29, 20, 5],
      [45,  6, 33, 35, 49, 25, 30, 47, 32, 31],
      [9, 12, 41, 15, 38, 14, 53, 22, 12, 13],
      [21, 24, 32, 49, 5, 47, 30, 23, 37, 8],
      [12, 19,  5, 28, 47, 39, 15, 35, 9, 51]]

SupplyCenter = ['S_1', 'S_2', 'S_3', 'S_4', 'S_5', 'S_6', 'S_7']
Supply = [110, 80, 100, 240, 100, 280, 130]
DemandMarket = ['D_1', 'D_2', 'D_3', 'D_4', 'D_5', 'D_6', 'D_7', 'D_8', 'D_9', 'D_10']
Demand = [90, 100, 150, 190, 180, 240, 210, 90, 160, 70]

import gurobipy as gb
from gurobipy import * 

S= len(SupplyCenter)
D= len(DemandMarket)


prob_9 = gb.Model("Problem 7")
P = prob_9.addVars(S,D, lb = 0, vtype = GRB.CONTINUOUS, name = [i+"_to_"+j for i in SupplyCenter for j in DemandMarket])

prob_9.setObjective(sum(C3[i][j]*P[i,j] for i in range(S) for j in range(D)) + 70 * (sum(Demand[j] for j in range(D)) - sum(P[i,j] for i in range(S) for j in range(D))),
                  GRB.MINIMIZE)


for i in range(S):
        prob_9.addConstr(sum(P[i,j] for j in range(D))==Supply[i])

for j in range(D):
    prob_9.addConstr(sum(P[i,j] for i in range(S))<=Demand[j])    


prob_9.optimize()

print("\nThe optimal allocation is:")
for v in prob_9.getVars():
    if v.x>0:
        print(v.varName, "=", round(v.x))
        
print("\nThe minimum cost for this problem is: $",prob_9.Objval)

print('Sensitivity Analysis (SA)\n ObjVal =', prob_9.ObjVal)

prob_9.printAttr(['Sense', 'Slack', 'Pi', 'RHS', 'SARHSLow', 'SARHSUp'])

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 17 rows, 70 columns and 140 nonzeros
Model fingerprint: 0xe2f22c2d
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [2e+01, 6e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e+01, 3e+02]
Presolve time: 0.01s
Presolved: 17 rows, 70 columns, 140 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0   -3.1000000e+33   1.400000e+32   3.100000e+03      0s
      16    4.1470000e+04   0.000000e+00   0.000000e+00      0s

Solved in 16 iterations and 0.02 seconds (0.00 work units)
Optimal objective  4.147000000e+04

The optimal allocation is:
S_1_to_D_9 = 110
S_2_to_D_3 = 60
S_2_to_D_6 = 20
S_3_to_D_6 = 80
S_3_to_D_7 = 20
S_4_to_D_2 = 100
S_4_to_D_6 = 140
S_5_to_D_1 = 90
S_5_to_D_9 = 10
S_6_to_D_5 = 180
S_6_to_D_8 = 30
S_6_to_D_10 = 70
S_7_to_D_3 = 90
S_7_to_D_9 = 40

The

From the above sensitivity analysis, we can deduce that the 3rd supplier (R2) is the best option of suppliers. Its shadow price is the smallest (-59) with a possible maximum value of 290, which is equilavent to an increase of up to 190 tonnes. Thus, it is the only supplier that can add 100 tonnes of milk to its supplying capacity.
Due to this boost, the change of cost is as follows:

41470-59 * 100+1000 = 41470 - 36570 = 4900

So the change of cost is of $4900.
